pennylane/templates/subroutines/uccsd_unitary.py

```diff
@@ -0,0 +1,231 @@
+
+    .. math::
+
+        \hat{U}(\vec{\theta}) =
+        \prod_{p > r} \mathrm{exp} \Big\{\theta_{pr}
+        (\hat{c}_p^\dagger \hat{c}_r-\mathrm{H.c.}) \Big\}
+        \prod_{p > q > r > s} \mathrm{exp} \Big\{\theta_{pqrs}
+        (\hat{c}_p^\dagger \hat{c}_q^\dagger \hat{c}_r \hat{c}_s-\mathrm{H.c.}) \Big\}
+
+    where :math:`\hat{c}` and :math:`\hat{c}^\dagger` are the fermionic annihilation and
+    creation operators and the indices :math:`r, s` and :math:`p, q` run over the occupied and
+    unoccupied molecular orbitals, respectively. Using the `Jordan-Wigner transformation
+    <https://arxiv.org/abs/1208.5986>`_ the UCCSD unitary defined above can be written in terms
+    of Pauli matrices as follows (for more details see
+    `arXiv:1805.04340 <https://arxiv.org/abs/1805.04340>`_):
+
+    .. math::
+
+        \hat{U}(\vec{\theta}) = && \prod_{p > r} \mathrm{exp} \Big\{ \frac{i\theta_{pr}}{2}
+        \bigotimes_{a=r+1}^{p-1} \hat{Z}_a (\hat{Y}_r \hat{X}_p - \mathrm{H.c.}) \Big\} \\
+        && \times \prod_{p > q > r > s} \mathrm{exp} \Big\{ \frac{i\theta_{pqrs}}{8}
+        \bigotimes_{b=s+1}^{r-1} \hat{Z}_b \bigotimes_{a=q+1}^{p-1}
+        \hat{Z}_a (\hat{X}_s \hat{X}_r \hat{Y}_q \hat{X}_p +
+        \hat{Y}_s \hat{X}_r \hat{Y}_q \hat{Y}_p +
+        \hat{X}_s \hat{Y}_r \hat{Y}_q \hat{Y}_p +
+        \hat{X}_s \hat{X}_r \hat{X}_q \hat{Y}_p -
+        \{\mathrm{H.c.}\}) \Big\}.
+
+    Args:
+        weights (array): length ``len(ph)+len(pphh)`` vector containing the parameters
+            :math:`\theta_{pr}` and :math:`\theta_{pqrs}` entering the Z rotation in
+            :func:`SingleExcitationUnitary <pennylane.templates.subroutine.SingleExcitationUnitary>`
+            and
+            :func:`DoubleExcitationUnitary <pennylane.templates.subroutine.DoubleExcitationUnitary>`.
+            These parameters are precisely the coupled-cluster amplitudes that need to be optimized
+            for each configuration generated by the excitation operator.
+        wires (Sequence[int]): wires the UCCSD unitary should act on
+        ph (Sequence[list]): sequence of two-element lists containing the indices ``r, p`` that
+            define the 1particle-1hole (ph) configuration :math:`\vert \mathrm{ph} \rangle =
+            \hat{c}_p^\dagger \hat{c}_r \vert \mathrm{HF} \rangle`,
+            where :math:`\vert \mathrm{HF} \rangle` denotes the Hartee-Fock (HF) reference state.
+        pphh (Sequence[list]): sequence of four-elements lists containing the indices
+            ``s, r, q, p`` that define the 2particle-2hole configuration (pphh)
+            :math:`\vert \mathrm{pphh} \rangle = \hat{c}_p^\dagger \hat{c}_q^\dagger \hat{c}_r
+            \hat{c}_s \vert \mathrm{HF} \rangle`.
+        init_state (array[int]): lenght ``len(wires)`` occupation-number vector representing the
+            HF state. ``init_state`` is used to initialize the qubit register.
+
```

Comment:
Question: we have the details `ph=None, pphh=None, init_state=None`. What happens if these are not defined, e.g., what does `None` correspond to in each case?

Reply:
Good point, I was also thinking about that.
- Probably, the best default for ``ph`` and ``pphh`` are empty lists. In the case both lists are not defined the template will just keep the HF state which does not make much sense. That's why I check for that in the template function.
- In the case of ``init_state`` is less clear as it depends on the number of electrons. One possibility might be taking ``init_state`` out of the template but that will make things more difficult for the user and the UCCSD template less complete.
What do you think?.